docs/internal/constrained-memory.md

```diff
@@ -0,0 +1,175 @@
+
+Most kernels provide two variants of memory allocation, one that can fail and one that can block indefinitely.
+For example, BSD's [`malloc(9)`](https://www.freebsd.org/cgi/man.cgi?query=malloc&sektion=9) requires either a `M_NOWAIT` or `M_WAITOK` flag to indicate whether blocking is allowed, Linux's [`kmalloc`](https://www.kernel.org/doc/htmldocs/kernel-api/API-kmalloc.html) provides a `GFP_NOWAIT` flag to disallow sleeping.
+This dichotomy is necessary because code holding non-sleepable kernel locks must not block on allocation failure.
+
+This exact problem should not apply in Verona because the memory allocator sits at the same abstraction layer as the rest of the runtime and strictly below any Verona code.
+This means that the memory allocator cannot depend on any cowns in the Verona world.
+There is a closely related problem: Verona behaviours are cooperatively scheduled.
+Blocking for an extended period waiting for memory can prevent the scheduler thread from making forward progress and so if there are as many behaviours blocked attempting to get memory as there are threads then the entire program will deadlock.
+
+Correctly handling allocation failure in Verona is difficult because the runtime itself needs to allocate memory for any `when` clause and so any response to out-of-memory (a global property) must be local and must not involve explicit or implicit allocation.
+Itanium C++ ABI runtimes are required, for example, to pre-allocate the memory for out-of-memory exceptions and to be able to throw an exception without allocation.
+This allows at least some threads to make forward progress, as long as they can recover without allocating any memory.
+
+### What do you do when you can't allocate?
+
+When a kernel runs out of memory, typically the first thing that it will try to do is discard memory from caches.
+A typical modern virtual memory subsystem has a set of 'clean' pages: ones that have been read from some backing store (e.g. an `mmap`ed file) and not modified.
+These can be discarded immediately.
+Note that this is a global property, not necessarily visible to the thread that cause memory exhaustion.
+
+The next thing that it will try to do is flush some things to the backing store.
+In a system with a unified buffer cache, there is no difference between anonymous memory backed by swap and memory-mapped files.
+Both can be written back to some backing store to free physical memory.
+This is somewhat more complex for the kind of deployment where we expect to see Verona: cloud services that may not have local storage.
+Swapping in general often requires some allocation, swapping to a remote blob store by sending some network messages may require a lot of memory, especially given that everything that is swapped will typically require encrypting.
+
+If this still fails, the kernel will often look for things to kill.
+Linux, for example, identifies the process with the most unsaved state of value to the user and kills it.
```

Comment:
While not, exactly, wrong...

Reply:
Is that not what their algorithm does?

Reply:
It is what their algorithm achieves, perhaps.